api/queue/work/model_service_deployment.go

```diff
@@ -92,18 +92,35 @@ func (depl *ModelServiceDeployment) Deploy(job *queue.Job) error {
 		isRedeployment = true
 	}
 
-	log.Infof("creating deployment for model %s version %s revision %s with endpoint id: %s", model.Name, endpoint.VersionID, endpoint.RevisionID, endpoint.ID)
+	// check if the latest deployment entry in the deployments table is in the 'pending' state (aborted workflow)
+	deployment, err := depl.DeploymentStorage.GetLatestDeployment(model.ID, endpoint.VersionID)
+	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+		log.Errorf("failed retrieving from db the latest deployment with the error: %v", err)
+		return err
+	}
 
-	// record the deployment process
-	deployment, err := depl.DeploymentStorage.Save(&models.Deployment{
-		ProjectID:         model.ProjectID,
-		VersionModelID:    model.ID,
-		VersionID:         endpoint.VersionID,
-		VersionEndpointID: endpoint.ID,
-		Status:            models.EndpointPending,
-	})
-	if err != nil {
-		log.Warnf("unable to create deployment history", err)
+	// create a new entry if the there is no deployment entry found or if it is in the 'pending' state
+	if deployment == nil || deployment.Status != models.EndpointPending {
```

Comment:
what is the `deployment.Status != models.EndpointPending` for?
Are there any case where deployment != nil and `deployment.Status != models.EndpointPending` by the fact it got into the job queue?

Reply:
maybe its easier to read if its just swap around (i know its the same) 😅 

When the last deployment is pending
```
if deployment != nil && deployment.Status == models.EndpointPending {
```

Reply:
+1 on swapping the conditional

Reply:
> what is the `deployment.Status != models.EndpointPending` for?

It's to check if the last entry in the deployments table has the deployment status set as pending (the deployment was aborted) or otherwise (in some terminal state like failed or deployed). In this case, if the status is a terminal state then we can safely create a new deployment entry in the table.

> Are there any case where `deployment != nil and deployment.Status != models.EndpointPending` by the fact it got into the job queue?

Uh yes, technically the deployment isn't what that gets into the job queue; a job gets into the job queue 😅 So if a job gets into a job queue, and if the last deployment entry fulfils the following condition, `deployment != nil and deployment.Status != models.EndpointPending`, that means that the last deployment is in a terminal state (deployed or failed).

> Are there any possibility that last deployment is in pending but there is no deployment ongoing? How will it be handled

Isn't this is supposed to be handled by the changes in this PR? Maybe I don't get what you're trying to ask 😅

> maybe its easier to read if its just swap around (i know its the same) 😅
When the last deployment is pending
`if deployment != nil && deployment.Status == models.EndpointPending {`

Yes I think the confusion might've stemmed from the fact that I chose to express the conditional in the inverse instead of the much more intuitive and natural alternative `if deployment != nil && deployment.Status == models.EndpointPending` (sign that I haven't coded in a long time perhaps 😅). Just so if it wasn't clear, `deployment == nil` happens when the model version is deployed for the first time.

Reply:
the confusion aside, just for my conceptual understanding
1. Merlin API Starts up, the jobqueue get populated with incompleted jobs
2. Incomplete jobs are processed by workers and `deployed` by this service
3. Hence`deployment.Status == models.EndpointPending` will always be true, since they are incompleted?

Reply:
Ahaha no worries, but yes, that'll always be true because the previous worker that was deploying the model wouldn't have updated the DB's deployments table to change the deployment's status from `pending` to `failed`/`deployed`.